xilem_core/src/sequence.rs

```diff
@@ -1,363 +1,651 @@
+///    sequence if the value is `Some`, and removed once the value is `None`.
+///  - A [`Vec`] of `ViewSequence` values.
+///    Note that this will have persistent allocation with size proportional
+///    to the *longest* `Vec` which is ever provided in the View tree, as this
+///    uses a generational indexing scheme.
+///  - Tuples of `ViewSequences` with up to 15 elements.
+///    These can be nested if an ad-hoc sequence of more than 15 sequences is needed.
+///
+pub trait ViewSequence<State, Action, Context: ViewPathTracker, Element: ViewElement, Marker>:
+    'static
+{
+    type SeqState;
+
+    /// Build the associated widgets into `elements` and initialize all states.
+    #[must_use]
+    fn seq_build(&self, ctx: &mut Context, elements: &mut AppendVec<Element>) -> Self::SeqState;
+
+    /// Update the associated widgets.
+    fn seq_rebuild(
+        &self,
+        prev: &Self,
+        seq_state: &mut Self::SeqState,
+        ctx: &mut Context,
+        elements: &mut impl ElementSplice<Element>,
+    );
+
+    /// Update the associated widgets.
+    fn seq_teardown(
+        &self,
+        seq_state: &mut Self::SeqState,
+        ctx: &mut Context,
+        elements: &mut impl ElementSplice<Element>,
+    );
+
+    /// Propagate a message.
+    ///
+    /// Handle a message, propagating to elements if needed. Here, `id_path` is a slice
+    /// of ids beginning at an element of this view_sequence.
+    fn seq_message(
+        &self,
+        seq_state: &mut Self::SeqState,
+        id_path: &[ViewId],
+        message: DynMessage,
+        app_state: &mut State,
+    ) -> MessageResult<Action>;
+}
 
-            fn mutate(&mut self, _cx: &mut $cx) -> &mut $pod
-            {
-                self.mutate()
-            }
+/// A temporary "splice" to add, update and delete in an (ordered) sequence of elements.
+/// It is mainly intended for view sequences.
+pub trait ElementSplice<Element: ViewElement> {
+    fn with_scratch<R>(&mut self, f: impl FnOnce(&mut AppendVec<Element>) -> R) -> R;
+    /// Insert a new element at the current index in the resulting collection.
+    fn push(&mut self, element: Element);
+    /// Mutate the next existing element.
+    fn mutate<R>(&mut self, f: impl FnOnce(Element::Mut<'_>) -> R) -> R;
+    /// Don't make any changes to the next n existing elements.
+    fn skip(&mut self, n: usize);
+    /// Delete the next existing element, after running a function on it.
+    fn delete<R>(&mut self, f: impl FnOnce(Element::Mut<'_>) -> R) -> R;
+}
 
-            fn mark(&mut self, changeflags: $changeflags, _cx: &mut $cx) -> $changeflags
-            {
-                self.last_mutated_mut().map(|pod| pod.mark(changeflags)).unwrap_or_default()
-            }
+/// Marker type to workaround trait ambiguity.
+#[doc(hidden)]
+pub struct WasAView;
+
+impl<State, Action, Context, V, Element> ViewSequence<State, Action, Context, Element, WasAView>
```

Comment:
```suggestion
impl<State, Action, Context, V, Element> ViewSequence<State, Action, Context, Element, fn() -> WasAView>
```

Otherwise it's e.g. not possible to `Clone` a ViewSequence (I just ran into that).
Alternatively we could also do
```rust
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default, Copy)]
pub struct WasAView;
```

Reply:
Oh, so it's a [perfect derive](https://smallcultfollowing.com/babysteps//blog/2022/04/12/implied-bounds-and-perfect-derive/) issue?
I'd say that you should just implement `Clone` for `Flex` manually; it's a logical mistake that `Flex: Clone` would require `Marker: Clone`, since that implementation is never used - working around it here isn't very principled.

That the same issue arises with `Debug` is unfortunate, and was a hazard I didn't see before. Overall, I'm happy to add the `Debug` derive here, but I'm going to push back on the fixes to make `Clone` work like that here.